UM/Signal.py

```diff
@@ -285,6 +286,63 @@ def __performEmit(self, *args, **kwargs):
 def strMethodSet(method_set):
     return "{" + ", ".join([str(m) for m in method_set]) + "}"
 
+##  A context manager that allows postponing of signal emissions
+#
+#   This context manager will collect any calls to emit() made for the provided signals
+#   and only emit them after exiting. This ensures more batched processing of signals.
+#
+#   The optional "compress" argument will limit the emit calls to 1. This means that
+#   when a bunch of calls are made to the signal's emit() method, only the last call
+#   will be emitted on exit.
+#
+#   \warning When compress is True, only the **last** call will be emitted. This means
+#   that any other calls will be ignored, _including their arguments_.
+#
+#   \param signals The signals to postpone emits for.
+#   \param compress Whether to enable compression of emits or not.
+@contextlib.contextmanager
+def postponeSignals(*signals, compress = False):
+    # Function that should replace the emit() method when compress == True
+    def postponedEmitCompressed(signal, *args, **kwargs):
+        signal._postponed = (args, kwargs)
+
+    # Function that should replace the emit() method when compress == False
+    def postponedEmit(signal, *args, **kwargs):
+        if not hasattr(signal, "_postponed"):
+            signal._postponed = []
+        signal._postponed.append((args, kwargs))
+
+    # To allow for nested postpones on the same signals, we should check if we actually made changes
+    # to the signal's emit method. If not, we can ignore it.
+    restore_emit = []
+    for signal in signals:
+        if not hasattr(signal, "_old_emit"): # Do nothing if the signal has already been changed
+            # Store the method for later retrieval
+            signal._old_emit = signal.emit
+
+            # Assign the relevant function, making sure it is seen as an instance method by Python.
+            if compress:
+                signal.emit = postponedEmitCompressed.__get__(signal, Signal)
+            else:
+                signal.emit = postponedEmit.__get__(signal, Signal)
+
+            # Since we made changes, make sure to restore the signal after exiting the context manager
+            restore_emit.append(signal)
+
+    yield
+
+    for signal in restore_emit:
+        signal.emit = signal._old_emit
+        delattr(signal, "_old_emit")
+        if hasattr(signal, "_postponed"):
+            if compress:
+                signal.emit(*signal._postponed[0], **signal._postponed[1])
+            else:
+                for args, kwargs in signal._postponed:
+                    signal.emit(*args, **kwargs)
+            delattr(signal, "_postponed")
+
+
```

Comment:
* I have moved the catching of signals into the emit() method. Note that I am using protected members directly in the context manager as I consider the context manager to be the "public API" for this functionality.
* I added a warning when emit() is called from a different thread than the postponing thread. Note though that it does not necessarily cause issues, it just means the emit() will be called later from a different thread.
* Good point, I have disabled connect/disconnect when a signals is being postponed.

As for the rest, I think having the postpone functionality is useful regardless of how we handle the signals elsewhere. Qt features something similar: http://doc.qt.io/qt-5/qobject.html#blockSignals and I have run into situations where it is really useful.